### ISSUE TITLE
`object-property-newline` should add the same string/object options as `object-curly-newline`

### ISSUE THREAD

**Issue by ezracelli** [eslint/eslint#11264]
**What rule do you want to change?**

`object-property-newline`

**Does this change cause the rule to produce more or fewer warnings?**

With the default config, the same

**What does the rule currently do?\
How will the change be implemented? (New option, new default behavior, etc.)?**

New options. Currently, the rule does not validate/recognize object destructuring patterns. The default configuration for this rule (`{ allowAllPropertiesOnSameLine: false }`) is essentially equivalent to the following `object-curly-newline` config:

```js
{
  ObjectExpression: 'always',
  ObjectPattern: 'never',
  ImportDeclaration: 'never'
  ExportDeclaration: 'always',
}
```

With the addition of object destructuring pattern recognition, I propose the default configuration should be simply `'always'`, aka:

```js
{
  ObjectExpression: 'always',
  ObjectPattern: 'always',
  ImportDeclaration: 'always'
  ExportDeclaration: 'always',
}
```

To my understanding, the customized config `{ allowAllPropertiesOnSameLine: true }` is able to be represented with the config `{ consistent: true }`, so the `allowAllPropertiesOnSameLine` option should be deprecated.

Overall, either globally or for each of the above four patterns, the options would be:

- String options `'always'` and `'never'`
- Object options `{ multiline: Boolean }`, `{ minProperties: Number }`, `{ consistent: Boolean }`, and `{ allowAllPropertiesOnSameLine: Boolean }` (with the latter-most deprecated)

**What will the rule do after it's changed?\
Please provide some example code that this change will affect:**

With the default config:

```js
/* eslint object-curly-newline: [ 'error', 'always' ] */
/* eslint-env es6 */

// object definition, still valid
let obj = {
  foo: 'foo',
  bar: 'bar',
  baz: 'baz',
}

// object definition, still invalid
obj = { foo: 'foo', bar: 'bar', baz: 'baz' }

// object pattern, still valid
let {
  foo,
  bar,
  baz,
} = obj

// object pattern, NEWLY INVALID
{ foo, bar, baz } = obj

//import declaration, still valid
import {
  foo,
  bar,
  baz,
} from 'file'

// import declaration, NEWLY INVALID
import { foo, bar, baz } from 'file'

// export declaration, still valid
export {
  foo,
  bar,
  baz,
}

// export declaration, still invalid
export { foo, bar, baz }
```

Equivalent to `{ allowAllPropertiesOnSameLine: true }`:

```js
/* eslint object-curly-newline: [ 'error', { consistent: true } ] */
/* eslint-env es6 */

// object definitions, still valid
let obj = { foo: 'foo', bar: 'bar', baz: 'baz' }
obj = {
  foo: 'foo',
  bar: 'bar',
  baz: 'baz',
}

// object definitions, still invalid
obj = {
  foo: 'foo', bar: 'bar',
  baz: 'baz',
}

// object patterns, still valid
let { foo, bar, baz } = obj
{
  foo,
  bar,
  baz,
} = obj

// object patterns, NEWLY INVALID
{
  foo, bar,
  baz
} = obj

// import declarations, still valid
import { foo, bar, baz } from 'file'
import {
  foo,
  bar,
  baz,
} from 'file'

// import declarations, NEWLY INVALID
import {
  foo, bar,
  baz,
} from 'file'
  

// export declarations, still valid
export default { foo, bar, baz }
export default {
  foo,
  bar,
  baz,
}

// export definitions, still invalid
export default {
  foo, bar,
  baz,
}
```

I won't duplicate the documentation for `object-property-newline`, but the additional options would be identical.

**Are you willing to submit a pull request to implement this change?**

Willing to give it a try!

**Comment by mattrabe** [comment]
For what it's worth I'd like to give my +1 👍 on this closed request. I would love to see this feature. I'm commenting to both encourage commitment at a future date if possible, and also get myself tagged on the thread so if it re-opens I may be of service on a PR, etc
